src/compas_ags/ags/constraints.py

```diff
@@ -0,0 +1,372 @@
+        constraint_jac_row = np.zeros((1, self.number_of_cols))
+
+        idx = self.vertex_index[self.vertex]
+        constraint_jac_row[0, idx] = 1 * math.sin(math.radians(self.angle))
+        r = (self.form.vertex_attribute(self.vertex, 'x') - self.x) * math.sin(math.radians(self.angle))
+
+        idx = self.vertex_index[self.vertex] + self.vcount
+        constraint_jac_row[0, idx] = 1 * math.cos(math.radians(self.angle))
+        r = (self.form.vertex_attribute(self.vertex, 'y') - self.y) * math.cos(math.radians(self.angle))
+        return constraint_jac_row, r
+
+    def update_constraint_goal(self):
+        self.set_initial_position()
+
+    def get_lines(self):
+        constraint_lines = []
+        s = self.form.vertex_coordinates(self.vertex, 'xy')
+        e = self.form.vertex_coordinates(self.vertex, 'xy')
+        theta = math.radians(90 - self.angle)
+        s[0] += 1 * math.sin(theta)
+        s[1] -= 1 * math.cos(theta)
+        e[0] -= 1 * math.sin(theta)
+        e[1] += 1 * math.cos(theta)
+        constraint_lines.append({
+            'start': s,
+            'end': e,
+            'width': self._width,
+            'color': self._color,
+            'style': self._style,
+        })
+        return constraint_lines
+
+
+class LengthFix(AbstractConstraint):
+    """Constraint that keeps the edge length fixed.
+
+    Parameters
+    -----------
+    form: :class:`FormDiagram`
+        The Form Diagram to constraint
+    edge: int
+        Index of the edge.
+
+    """
+
+    def __init__(self, form, edge):
+        super().__init__(form)
+        self.edge = edge
+        self.L = None
+        self.set_initial_length()
+
+    def update_constraint_goal(self):
+        self.set_initial_length()
+
+    def set_initial_length(self):
+        u, v = list(self.form.edges())[self.edge]
+        s = self.form.vertex_coordinates(u, 'xy')
+        e = self.form.vertex_coordinates(v, 'xy')
+        dx = s[0] - e[0]
+        dy = s[1] - e[1]
+        self.L = math.sqrt(dx ** 2 + dy ** 2)  # Initial length
+
+    def compute_constraint(self):
+        constraint_jac_row = np.zeros((1, self.number_of_cols))
+
+        u, v = list(self.form.edges())[self.edge]
+        s = self.form.vertex_coordinates(u, 'xy')
+        e = self.form.vertex_coordinates(v, 'xy')
+        dx = s[0] - e[0]
+        dy = s[1] - e[1]
+        length = math.sqrt(dx ** 2 + dy ** 2)  # Current length
+
+        constraint_jac_row[0, u] = dx / length  # x0
+        constraint_jac_row[0, v] = -dx / length  # x1
+        constraint_jac_row[0, u + self.vcount] = dy / length  # y0
+        constraint_jac_row[0, v + self.vcount] = -dy / length  # y1
+        r = length - self.L
+
+        return constraint_jac_row, r
+
+
+class SetLength(LengthFix):
+    """Constraint that sets the edge length to L.
+
+    Parameters
+    -----------
+    form: :class:`FormDiagram`
+        The Form Diagram to constraint
+    edge: int
+        Index of the edge.
+    L: float
```

Comment:
not a fan of `L` and `P` since we tend to use capital letters for matrices...

Reply:
why not just length

Reply:
changed to length and either `self.x` or `self.y` I think it is consistent.